### ISSUE TITLE
FT-1 — Heads-as-modules: uniform head interface, replay assembly, FPS subselection, LoRA and freezing

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1580]
**Depends on:** TRN-1 (#1575), ARCH-3 (#1562) · **Blocks:** FT-2 (#1581), FT-3 (#1582), FT-4 (#1637), GATE-3 (#1585)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy fine-tuning is scattered special cases smeared across three files. `mace/tools/multihead_tools.py` (494 lines) mixes config, data, and runtime state: `HeadConfig` (`:30`) is a dataclass of 21 fields, all but two `Optional`, where `train_loader`, `z_table`, `collections`, `mean`/`std` are filled in later as scratch state; `prepare_pt_head` (`:103`) special-cases the curated dataset names `mp`/`omat`/`matpes_pbe`/`matpes_r2scan`; `assemble_replay_data` (`:146`) downloads a curated XYZ from the `ACEsuit/mace-foundations` GitHub releases, runs subselection, writes an intermediate `mp_finetuning-<tag>.xyz` to the working directory and re-reads it (`:182,:202-204`). Head prep is welded into the legacy `run()` (`mace/cli/run_train.py:275-322`), including the replay-vs-real data-ratio upsampling at `run_train.py:444-459`. Subselection lives in a separate 575-line CLI, `mace/cli/fine_tuning_select.py` (`mace_finetuning_select`), which per the migration disposition MERGEs into the integrated pipeline (its CLI flags `--configs_pt --configs_ft --num_samples --subselect --model --output --descriptors --device --default_dtype --head_pt --head_ft --filtering_type --weight_ft --weight_pt --filter_atomic_numbers_pt --disallow_random_padding --seed` map into it). LoRA lives in `mace/modules/lora.py` (356 lines: `LoRAO3Linear` wrapping e3nn's `o3.Linear` at `:30`, `LoRADenseLinear` `:163`, `LoRAFCLayer` `:219`, `inject_lora` `:287`, `inject_LoRAs` `:325`, `merge_lora_weights` `:330`), injected at `run_train.py:842-862` and merged before save at `:1139-1141`. Freezing is `--freeze` inside `get_params_options` (`mace/tools/scripts_utils.py:930`, thresholds at `:944-960`). This ticket redesigns all of it around one principle (the fine-tuning axis of the redesign): **the low-level fitting interface treats all heads the same — replay vs real head differ only in the source of configurations and labels.** Behavior is pinned by the P0-5 fine-tuning contract (new syntax, same semantics); legacy is read in place as the oracle, never edited.

**Interface & constraints:**

1. **Uniform head interface.** A head is a typed spec — name, data source, label source, `KeySpecification`, per-head weighting, per-head `E0Spec` — with **no runtime state** (the 21-field `Optional`-soup `HeadConfig` split into immutable config vs stage-produced state; no argparse `Namespace` crosses the interface — legacy passes one into `dict_head_to_dataclass` and `prepare_pt_head`, `multihead_tools.py:53,:103`). The replay head (`pt_head`) is the **same class/protocol** as a real head, constructed with a different source; nothing downstream may branch on "is this the replay head" — legacy branches on the literal string in at least four places (`run_train.py:453-455`, `:559`, `:641`, and the head sort key at `:572`). Each head's data source yields `list[Configuration]` + provenance — "replay vs real differ only in the source" must be literally true at the interface. Torch-free spec types live in `mace_core`; implementations in `mace_torch`. Heads flow through TRN-1 (#1575)'s `DataStage` like any other head; head/readout *creation* from declared observables is ARCH-3 (#1562)'s job — FT-1 (#1580) supplies the data/label-source layer, the adapters, and the freezing.

2. **Replay head E0s are fixed** to the typed spec `E0sFromFoundation(missing="error")` (the E0Spec union, resolved once before model build by ARCH-3 (#1562)/TRN-1 (#1575)): a species missing from the foundation model's E0 table is a **hard error**, never the `head_energies.get(z, 0.0)` zero-pad legacy applies under `--foundation_model_elements` (`run_train.py:578-586`). Reading the FM's E0s from `model_foundation.atomic_energies_fn.atomic_energies` by attribute reflection (`run_train.py:509-519,:560-571`) does not carry over either — v1 reads the artifact's `e0s` metadata block.

3. **Replay data assembly** (port of `assemble_replay_data`): curated name → release-asset URL, downloaded once and cached under `~/.cache/mace` (`XDG_CACHE_HOME` override — `get_cache_dir`, `mace/tools/utils.py:170-172`), then subselected and streamed as `Configuration`s — no intermediate XYZ round-trip through the working directory. Curated names and sources (from `multihead_tools.py:154-165`, all four under `github.com/ACEsuit/mace-foundations/releases/download/`): `mp` → `mace_mp_0b/mp_traj_combined.xyz`, `omat` → `mace_omat_0/mp_traj_combined_omat.xyz`, `matpes_pbe` → `mace_matpes_0/matpes-pbe-replay-data.xyz`, `matpes_r2scan` → `mace_matpes_0/matpes-r2scan-replay-data.extxyz`. The replay-vs-real ratio guard keeps its semantics: when (real train size − replay size) / replay size falls below the configured threshold (legacy `--real_pt_data_ratio_threshold`, default 0.1), the fine-tuning heads are upsampled by the integer factor legacy computes at `run_train.py:445-459`. Legacy skips the guard entirely for LMDB/HDF5 heads (`:458-461`); v1 keeps that carve-out explicit rather than implicit in an `if all_ase_readable`.

4. **Subselection** (port of `fine_tuning_select.py` as a library function; CLI exposure is CLI-1 (#1579)'s): descriptor-based farthest-point sampling using foundation-model descriptors, with `SubselectType` ∈ {`fps`, `random`} (`fine_tuning_select.py:39`; `fpsample` optional dependency for fast FPS, with a fallback when absent) — note the two defaults differ and both must be preserved: the standalone CLI defaults to `fps` (`:116`) while `--subselect_pt` on `mace_run_train` defaults to `random` (`arg_parser.py:603-607`). `FilteringType` ∈ {`none`, `combinations`, `exclusive`, `inclusive`} (`fine_tuning_select.py:29`, semantics documented at `:209-217`), `num_samples`, seed, atomic-number filter, and the random-padding rule (legacy `--disallow_random_padding_pt` is a `store_false` onto `dest="allow_random_padding_pt"`, `arg_parser.py:615-619`, so the *default is to allow* padding — v1 exposes one positive boolean, not the negated flag). Selection behavior is a flagged pinning gap: the four `tests/workflows/test_finetuning_select.py` cases are the behavioral spec — same seed ⇒ same selected configurations.

5. **LoRA adapters** (defaults: rank 4, alpha 1.0 — `arg_parser.py:651-661`) attach to the v1 dispatched `linear` ops in their **canonical weight layout** (the BKD-1 (#1560) op contract) — the legacy `o3.Linear` wrappers are legacy-only and do not carry over (no e3nn in v1). Adapter parameters are the only trainables when LoRA is on; `merge_lora_weights` semantics are preserved: merging before checkpoint save produces an artifact with the **same shapes as a non-adapted model** (no LoRA trace), so it loads into any backend. Port the `tests/unit/test_lora.py` cases.

6. **Freezing** keeps the legacy layered semantics of `--freeze N` (`mace/tools/scripts_utils.py:944-960`): N ≥ 1 freezes the node embedding, ≥ 5 the interactions, ≥ 6 the products, ≥ 7 the readouts — each freeze sets `requires_grad=False` *and* zeroes that group's learning-rate factor in `lr_params_factors`. Expressed in the typed `finetune` config section (CFG-1 (#1574)), not by namespace mutation and not by JSON-in-a-string. Port both `tests/workflows/test_freeze.py` cases (`test_run_train_freeze`, `test_run_train_soft_freeze` — the "soft" case is `--lr_params_factors` without `--freeze`, and it must stay expressible).

7. **Preserved capability surface** (KEEP dispositions; new config names, same semantics): `--foundation_model` / `--foundation_model_kwargs` / `--foundation_model_readout`; `--multiheads_finetuning` / `--heads` / `--foundation_head`; `--weight_pt_head` / `--num_samples_pt` / `--real_pt_data_ratio_threshold`; `--pt_train_file` / `--pt_valid_file`; `--subselect_pt` / `--filter_type_pt` / `--disallow_random_padding_pt`; `--lora` / `--lora_rank` / `--lora_alpha`; `--freeze` / `--lr_params_factors`. Foundation-model presets (legacy overwrites `args.lr = 0.0001`, `args.ema = True`, `args.ema_decay = 0.99999` at `run_train.py:247-249` unless `--force_mh_ft_lr`) become **declared defaults in the resolved config** — visible, overridable, and with no `--force_mh_ft_lr` equivalent because nothing needs forcing.

**Task:**
1. `mace_core`: typed head spec (name, data source, label source, key spec, weighting, `E0Spec`) + provenance types; no torch, no runtime state.
2. `mace_torch`: real-head data source (files) and replay-head data source (curated download + cache + subselect), both yielding `list[Configuration]` + provenance; ratio upsampling guard with the explicit non-ASE carve-out.
3. Subselection library: FPS (`fpsample` + fallback) and random, all four filtering types, the positive padding boolean — behavior matched to the ported `test_finetuning_select.py` cases, with both legacy defaults (`fps` standalone, `random` in-run) preserved.
4. LoRA adapters over the v1 `linear` ops (canonical layout); inject / train-adapters-only / merge-on-save; port `test_lora.py` cases.
5. Freezing + per-group lr factors in the `finetune` config section; port both `test_freeze.py` cases.
6. Wire heads through TRN-1 (#1575)'s `DataStage`/`TrainStage`; run the P0-5 fine-tuning contract under `MACE_ENGINE=v1` (anchor-as-foundation, no network).

**Out of scope:** pseudolabel generation, fine-tune shortcuts/best-practice presets, replay-dependent defaults, and worked docs (FT-2 (#1581)); all-species weight saving, new-species initialization, stage-callable fine-tuning (FT-3 (#1582)); the MDP dipole/polarizability fine-tune mode (FT-4 (#1637)); the E0 resolver implementation (ARCH-3 (#1562)); foundation-model checkpoint conversion (FM-00 (#1565)/FM-1 (#1586)); CLI subcommand surface (CLI-1 (#1579)).

**Acceptance criteria:**
- [ ] Replay and real heads are the same class/protocol; grep-level check: no `pt_head`/replay `isinstance` or name-string branching anywhere downstream of the head interface (legacy has at least four such branches).
- [ ] The P0-5 fine-tuning contract passes under `MACE_ENGINE=v1`: multihead replay run from a P0-1 anchor-as-foundation completes, the model contains `pt_head` + the new head, and fine-tuning-set error decreases vs the unadapted model.
- [ ] Subselection reproduces legacy selection on the ported `test_finetuning_select.py` cases (same seed → same selected configs) for both `fps` and `random` and all four filtering types, including the padding allow/disallow rule and both legacy defaults.
- [ ] LoRA: with adapters on, only adapter parameters receive gradients; merged model reproduces the adapted model's outputs at fp64 tolerance; the merged checkpoint has non-adapted shapes and round-trips through the backend registry.
- [ ] Freeze thresholds map exactly (≥1 embedding, ≥5 interactions, ≥6 products, ≥7 readouts) with `requires_grad` flags and zeroed lr factors asserted; the `--lr_params_factors`-without-`--freeze` "soft freeze" path stays expressible (both ported `test_freeze.py` cases green).
- [ ] Replay ratio upsampling reproduces the legacy integer-factor semantics at the threshold boundary, and an LMDB/HDF5 head skips the guard by an explicit rule.
- [ ] A species absent from the foundation E0 table fails loudly at data-stage time (replay `E0sFromFoundation(missing="error")`); no zero-pad path exists, and FM E0s are read from artifact metadata, not by attribute reflection on a loaded module.
- [ ] No argparse namespace object crosses the head interface; head specs are immutable typed objects.

**Inventory gaps assigned here:**
- [ ] Fine-tuning selection behaviour (`--subselect_pt`, `--filter_type_pt`, `--disallow_random_padding_pt`) — the four `test_finetuning_select.py` cases exercise the library, nothing exercises the in-run `mace_run_train` path.
- [ ] `--foundation_model_kwargs` / `--foundation_model_readout` / `--foundation_head` — no test asserts a non-default value is honoured.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/finetune -v                      # heads, sources, subselect, lora, freeze units
MACE_ENGINE=v1 MACE_CI_ALLOW_NETWORK=1 python -m pytest tests/workflows/test_finetuning_pseudolabels.py -v   # replay contract (network-marked: pytestmark at :11)
MACE_ENGINE=v1 python -m pytest tests/workflows/test_finetuning_select.py tests/workflows/test_freeze.py -v  # ported selection + freeze behavior
python -m pytest tests/unit/test_lora.py tests/unit/test_multihead_tools.py -v
```
**Review focus:** that the head interface is genuinely uniform (the replay head must not be a special case anywhere), the LoRA merge numerics against the canonical layout, and selection reproducibility — fine-tuning semantics need reviewer familiarity with the foundation-model workflow.

---
